_docs/Configuring-Portus.md
---
title: Configuring Portus
layout: default
order: 1
longtitle: How to configure Portus
---

## The config.yml file

Before starting this application, you may want to configure some values that are
evaluated during the initialization process of Portus. All these values are
specified in the `config/config.yml` file. This file contains the default values
for each setting, and you should *never* touch it. If you want to modify some of
its values you have two options: creating a `config/config-local.yml` file with
the values that you override the default ones, or to use **environment
variables**. For example, imagine the following situation:

{% highlight yaml %}
# In config.yml
settings:
  a: true
  b: false

# In config-local.yml
settings:
  a: false
{% endhighlight %}

The result of the previous example is that both `a` and `b` are false.

In containerized deployments though, environment variables are usually more
convenient. Let's imagine the following configuration:

{% highlight yaml %}
feature:
  enabled: true
  value: "val"
{% endhighlight %}

In Portus we follow a **naming convention** for environment variables: first of
all we have the `PORTUS_` prefix, and then we add each key in uppercase. So, for
example, the previous example can be tweaked by setting:
`PORTUS_FEATURE_ENABLED` and `PORTUS_FEATURE_VALUE`.

Note that environment variables override even values from the `config-local.yml`
file. So, to sum this up, you can assume that Portus follows the following
preference when reading a configuration value (listed from max. preference to
least):

1. Environment variables.
2. The user-defined `config-local.yml` file.
3. The default values from the `config.yml` file.

Last but not least, some of these configuration values are quite delicate. This
is why you should manage them with secrets, [as explained
here](/docs/secrets.html).

## List of configuration options

### Email configuration

{% highlight yaml %}
email:
  from: "portus@example.com"
  name: "Portus"
  reply_to: ""

  smtp:
    enabled: false
    address: "smtp.example.com"
    port:    587,
    domain:  "example.com"

    ##
    # SSL.

    ssl_tls:              ""
    enable_starttls_auto: false
    openssl_verify_mode:  "none"
    ca_path:              ""
    ca_file:              ""

    ##
    # Authentication

    user_name:      ""
    password:       ""
    authentication: "login"
{% endhighlight %}

There are some important things to note from **SMTP** support:

1. If SMTP is disabled, then `sendmail` is used instead (the specific command
   being: `/usr/sbin/sendmail -i -t`).
2. Most options described here have a direct translation in
   [ActiveMailer::Base](https://api.rubyonrails.org/v5.1/classes/ActionMailer/Base.html),
   which is the mailer used by Portus. If you go over the "Configuration
   options" section, you will start to see the same keys as in the Portus
   configuration, and their documentation.
3. The `ssl_tls` key is a combination of the `ssl` and the `tls` options as
   described in the [documentation of
   ActiveMailer::Base](https://api.rubyonrails.org/v5.1/classes/ActionMailer/Base.html). If
   you don't want SSL support, then simply leave this option empty. Otherwise,
   you will have to write either "ssl" or "tls" to enable any of these
   options. Moreover, `openssl_verify_mode` accepts only two values: "none" (the
   default), or "peer". Again, refer to the same documentation for more info.
3. Authentication is only enabled if you provide a `user_name`. Moreover, note
   that the password that you have to write might be different than the one you
   have on your personal account (e.g. some providers like GMail ask you to
   create application passwords if you are using two-factor
   authentication). Last but not least, `authentication` can only have three
   values: "login", "plain" or "cram_md5". Check the [documentation of
   ActiveMailer::Base](https://api.rubyonrails.org/v5.1/classes/ActionMailer/Base.html)
   in order to see what's the right value for you.
4. Note that this configuration section will be directly mapped into options
   that are passed into the mailer. We are using the same mailer as many other
   Ruby on Rails applications, so all the examples posted on different places
   should also apply here. For example, Gitlab has written an [awesome
   list](https://docs.gitlab.com/omnibus/settings/smtp.html) which describes
   which options are to be used for each mail provider.

An example configuration for GMail might look like:

{% highlight yaml %}
email:
 from: "account@gmail.com"
 name: "Name"

 smtp:
   enabled: true
   address: "smtp.gmail.com"
   port: 587,
   domain: "smtp.gmail.com"
   user_name: "account@gmail.com"
   password: "two-factor-password"
   authentication: "login"
   enable_starttls_auto: true
   openssl_verify_mode:  "none"
{% endhighlight %}

### Gravatar

If enabled, then the profile picture will be picked from the [Gravatar](https://en.gravatar.com/) associated with each user. Otherwise, a default icon will be shown.

{% highlight yaml %}
gravatar:
  enabled: true
{% endhighlight %}

### Delete support

**Note**: bear in mind that this will only work accordingly if you are using
Docker Distribution 2.4 or later.

As of 2.1, and if you are using a Docker Distribution version not older than
2.4, you will be able to delete images, tags, users, namespaces and teams. In
order to do so just enable the `delete` option:

{% highlight yaml %}
delete:
  enabled: true
  contributors: false
  garbage_collector:
    enabled: false
    older_than: 30
    keep_latest: 5
    tag: ""
{% endhighlight %}

This option is **disabled** by default. This is because we want users enabling
this if they are really sure about the feature itself and its requirements. For
more information, read [this page](/features/removing_images.html).

Moreover, this action can only be performed by team owners at first. You can
change this by setting `contributors` to `true`, in which case contributors will
also be able to remove images and tags.

Last but not least, you can tell Portus to automatically remove tags which are
considered old. This is documented
[here](/features/removing_images.html#automating-the-removal-of-images-and-tags).

### LDAP Support

If enabled, then only users of the specified LDAP server will be able to use Portus.

{% highlight yaml %}
ldap:
  enabled: false

  hostname: "ldap_hostname"
  port: 389
  timeout: 5

  encryption:
    method: ""
    options:
      ca_file: ""
      ssl_version: "TLSv1_2"

  base: ""
  admin_base: ""
  filter: ""
  uid: "uid"

  authentication:
    enabled: false
    bind_dn: ""
    password: ""

  guess_email:
    enabled: false
    attr: ""
{% endhighlight %}

Some notes:

- **base**: The base where users are located
  (e.g. "ou=users,dc=example,dc=com").
- **admin_base**: The base where admin users are located.
  (e.g. "ou=admin,dc=example,dc=com"). Use this base if you have a separate
  route for admin users. Users logging in from this base will be considered
  Portus administrators automatically. Available since **Portus 2.5**.
- **filter**: This option comes in handy when you want to filter even further
  the results that might be hanging from the *base*.
- **encryption**
  - **method**: The method of encryption between Portus and the LDAP server. It
    defaults to an empty value, which means that the communication won't be
    encrypted. You can also use "simple_tls" and "starttls".
  - **options**: You can pass some extra options, like a path to a **ca_file**
    (by default it will use the parameters from the host), and the
    **ssl_version**.
- **guess_email**: Portus needs an email for each user, but there's no standard
  way to get that from LDAP servers. You can tell Portus how to get the email
  from users registered in the LDAP server with this configurable value.
- **uid**: The attribute where Portus will look for the user ID when
  authenticating.
- **authentication**: Some LDAP servers require a binding user in order to
  authenticate. You can specify this user by enabling this option. Then you
  should provide the DN of this user in the `bind_dn` value.

There are three possibilities for the **guess_email** option:

- disabled: this is the default value. It means that Portus won't do a thing
  when registering LDAP users (users will be redirected to their profile page
  until they setup an email account).
- enabled where "attr" is empty: for this you need "ldap.base" to have some
  value. In this case, the hostname will be guessed from the domain component of
  the provided base string. For example, for the dn:
  "ou=users,dc=example,dc=com", and a user named "mssola", the resulting email
  is "mssola@example.com".
- enabled where "attr" is not empty: with this you specify the attribute inside
  a LDIF record where the email is set.

If something goes wrong when trying to guess the email, then it just falls back
to the default behavior (empty email).

### Creating the first admin user

As explained [here](/features/3_teams_namespaces_and_users.html), an admin user
can upgrade users to be administrators. But, how do we create the very first
admin user? There are two ways in which this can be done:

{% highlight yaml %}
first_user_admin:
  enabled: true
{% endhighlight %}

By default the first user to be created will be granted admin permissions. In
this case, when you go to the "Sign up" page, you will find the following
situation:

![Creating the Admin](/images/docs/create-admin.png)

That is, the "Sign up" form is telling you that the user to be created will be
an admin. Therefore, in order to create the first admin user, you just have to
proceed as usual and fill the "Sign up" form for this user. After doing this,
the "Sign up" form will not be able to create administrators, but only regular
users.

However, note if LDAP support is enabled, the first user that logs in with LDAP
credentials will be automatically an admin. This is shown in the following
screenshot:

![Creating the Admin in LDAP](/images/docs/create-admin-ldap.png)

The other way to do this is to disable the `first_user_admin` configurable
value. In this case, the first admin cannot be created from the Web UI, rather
you have to use the `portus:make_admin` rake task. Therefore, you should
access into the Portus application and run:

    $ rake portus:make_admin[<username>]

When you are using the [official Docker
image](https://hub.docker.com/r/opensuse/portus/), you can do this with:

    $ docker exec -it <container-id> portusctl make_admin <username>

### Disabling the sign up form

By default Portus will have the sign up form available for any person that comes
across your Portus instance. You can change that by disabling the `signup`
configuration value:

{% highlight yaml %}
signup:
  enabled: false
{% endhighlight %}

This way, only the admins of this Portus instance will be able to create users
in the system. This setting is completely ignored when LDAP support is enabled.
This option is explained with more detail
[here](/features/disabling_signup.html).

### Enforce SSL usage on production

By default Portus will enforce usage of SSL when ran with the "production"
environment. This is required for security reasons. The behaviour is controller
by this configuration setting:

{% highlight yaml %}
check_ssl_usage:
  enabled: true
{% endhighlight %}

### OAuth support

**Note**: this is only available in Portus 2.3 or later.

Portus provides OAuth support: you can configure it to support authentication
from one of the supported platforms. Here is the full list:

{% highlight yaml %}
oauth:
  local_login:
    enabled: true

  google_oauth2:
    enabled: false
    id: ""
    secret: ""
    domain: ""
    options:
      # G Suite domain. If set, then only members of the domain can sign in/up.
      # If it's empty then any google users con sign in/up.
      hd: ""

  open_id:
    enabled: false
    identifier: ""
    domain: ""

  github:
    enabled: false
    client_id: ""
    client_secret: ""
    organization: ""
    team: ""
    domain: ""

  gitlab:
    enabled: false
    application_id: ""
    secret: ""
    group: ""
    domain: ""
    server: ""

  bitbucket:
    enabled: false
    key: ""
    secret: ""
    domain: ""
    options:
      team: ""
{% endhighlight %}

As you can see, the only option enabled by default is `local_login`, which tells
Portus that users can log in with their Portus credentials. You can disable this
if you want users to only login with OAuth credentials.

All supported platforms have their own settings, but there are some common
attributes that might not be evident:

- *domain*: if a domain (e.g. mycompany.com) is set, then only signups with
  emails from this domain are allowed.
- *organization*, *group* and *team*: this is used by providers like Github,
  Gitlab or Bitbucket. With this you can restrict the team or organization where
  the given user belongs.
- OpenID's *identifier* and Gitlab's *server*: these attributes enable Portus to
  fetch the proper source. That is, you can specify the Gitlab server (by
  default gitlab.com), or the OpenID provider (by default you'll be prompted
  with a form asking for this information).

For more info, take a look at the [dedicated page](/features/oauth.html) for
OAuth support.

### Advanced registry options

You can configure some aspects on how Portus interacts with your Docker registry:

{% highlight yaml %}
registry:
  jwt_expiration_time:
    value: 5

  catalog_page:
    value: 100

  timeout:
    value: 2

  read_timeout:
    value: 120
{% endhighlight %}

The JWT token is one of the main keys in the authentication process between
Portus and the registry. This token has as one of its values the expiration
time of itself. The problem is that the registry does not request another token
when it expires. This means that for large images, the upload might fail
because it takes longer than the expiration time. You can read more about this
in the issue [SUSE/Portus#510](https://github.com/SUSE/Portus/issues/510).

To workaround this, we allow the admin of Portus to raise the expiration time
as required through the `jwt_expiration_time` configurable value. This value is
set in minutes, but we still allow the syntax as allowed before the 2.1 release
(deprecated in 2.3 and to be removed in 2.4).

Moreover, the registry might be slow or the network connection flaky. For this,
you can also use the `timeout` and the `read_timeout` values. This has been
added in Portus 2.3.

Finally, another option is the `catalog_page`. This tweaks the page size for
each catalog request performed by Portus. The default value for this should be
enough for the vast majority of users, but with this we allow administrators to
workaround a Docker Registry bug as stated
[here](https://github.com/SUSE/Portus/issues/1001).

### FQDN of your machine

On the next release of Portus, the FQDN of the machine is no longer a secret
and it's now considered a configurable value. The fact that it was a secret
before is because of legacy code. However, you now can configure it like this:

{% highlight yaml %}
machine_fqdn:
  value: "portus.test.lan"
{% endhighlight %}

### Security scanning

**Note**: this is only available in Portus 2.3 or later.

You can setup a security scanner for your Portus instance. This way, Portus will
display whether the images you have on your registry have known security
vulnerabilities. This feature is disabled by default, but it can be configured
through the following values:

```yaml
security:
  # CoreOS Clair support (https://github.com/coreos/clair). This is only
  # guaranteed to work for v2.0.x releases of Clair.
  clair:
    server: ""
    health_port: 6061
    timeout: 900

  # zypper-docker can be run as a server with its `serve` command. This backend
  # fetches the information as given by zypper-docker. Note that this feature
  # from zypper-docker is experimental and only available through another branch
  # than master.
  #
  # NOTE: support for this is experimental since this functionality has not
  # been merged into master yet in zypper-docker.
  zypper:
    server: ""

  # This backend is only used for testing purposes, don't use it.
  dummy:
    server: ""
```

Portus supports having multiple scanners enabled at the same time. You can read
more about this [here](/features/6_security_scanning.html).

### Anonymous browsing

**Note**: feature only available in Portus 2.3 or later.

You can configure Portus to allow anonymous users to explore public
repositories. This feature is enabled by default:

{% highlight yaml %}
anonymous_browsing:
  enabled: true
{% endhighlight %}

You can read more about this [here](/features/anonymous_browsing.html).

### Display name

This option tells Portus to not use the username when showing users, rather a
user-defined "display name". You can read more about this
[here](/features/display_name.html). By default this feature is disabled, and
it's controlled with this setting:

{% highlight yaml %}
display_name:
  enabled: false
{% endhighlight %}

### Granular user permissions

You can further tweak the permissions that users have on a Portus instance with
the following options:

{% highlight yaml %}
user_permission:
  change_visibility:
    enabled: true

  create_team:
    enabled: true

  manage_team:
    enabled: true

  create_namespace:
    enabled: true

  manage_namespace:
    enabled: true

  create_webhook:
    enabled: true

  manage_webhook:
    enabled: true

  push_images:
    policy: allow-teams
{% endhighlight %}

- **change_visibility**: allow users to change the visibility or their personal
  namespace. If this is disabled, only an admin will be able to change this. It
  defaults to true.
- **create/manage_team**: allow users to create/modify teams if they are an
  owner of it. If this is disabled, only an admin will be able to do this. It
  defaults to true.
- **create/manage_namespace**: allow users to create/modify namespaces if they
  are an owner of it. If this is disabled, only an admin will be able to do
  this. It defaults to true.
- **create/manage_webhook**: allow users to create/modify webhooks if they are
  an owner of it. If this is disabled, only an admin will be able to do this. It
  defaults to true.
- **push_images**: set a push policy. Available options are: `allow-teams`,
  `allow-personal` and `admin-only`. You can read more about push policies
  [here](/features/3_teams_namespaces_and_users.html).

### Background process

**Note**: this is only available in Portus 2.3 or later.

As explained in the [documentation page](/docs/background.html) of the background
process, we have introduced a background process to alleviate some heavy
operations from the main Portus process. This background process has its own
configuration values:

{% highlight yaml %}
background:
  registry:
    enabled: true

  sync:
    enabled: true
    strategy: initial
{% endhighlight %}

- The `security scanning` task can be disabled by not providing a server as
  described [here](/docs/Configuring-Portus.html#security-scanning).
- The `registry` integration can be disabled, but it is **highly discouraged**
  since then all registry events will be missed by Portus (they will be logged,
  but not handled).
- The `sync` task can be disabled as well, but more importantly you can tune it
  further when enabled with the `strategy` option. This option may have any of
  the following values:
  - **update-delete**: it performs a full synchronization (traditional behavior
    of the old `crono` process).
  - **update**: it only adds missing tags, but it does not remove any contents
    from the database.
  - **on-start**: when starting Portus it runs an `update-delete` and then it
    gets disabled (i.e. it will only run once).
  - **initial**: like `on-start`, but it only runs if the database is
    empty. This is the default value since it's deemed to be the most common
    use-case and the safest option. The idea behind this option is that you only
    want to synchronize when bootstrapping your instance.

### Pagination


CHANGELOG.md
## Upcoming Version

## 2.4.1

This release consists mostly of bug fixes and upgrades from vulnerable
gems. Thank you everyone involved!

- af08f6211abb db: added index on scan_result
- a282fff71bd4 Properly show tokens for bots
- a55c0ec2f827 Upgraded cconfig to fix a configuration issue
- d3be549af55f Fixed namespace duplication
- 3591e04ba75c ui: fixed team creation for standard user
- 52915908e36b ui: fixed repositories performance view issue
- 69e41ece3f8a Reduced amount of rendered data for repository entity
- 942f18113fa1 Fixed breaking changes from pagination commit
- 762c9665bb11 policies: fixed destroy for repositories/tags
- Upgraded the following gems due to vulnerabilities:
  - 5f90273e0b9a nokogiri
  - 351fc7d27b3e loofah
  - d3be549af55f rack
  - ae62d4c4c0b7 rails

## 2.4.0

### Highlight

#### Configuration changes

We have introduced quite some configurable options. Some of them are new, and
some other are merely additions to existing ones.

First of all, we have expanded the configuration for the mailer. We are now
providing more options so administrators have more flexibility in regards to how
they manage SSL/TLS. You can read the update documentation of the mailer
[here](http://port.us.org/docs/Configuring-Portus.html#email-configuration).

Moreover, the `delete` option has now two new options:

1. You can allow contributors to delete namespaces/repositories/etc. with the
   `delete.contributors` option (it's set to `false` by default).
2. The background process can now automatically remove images that are older
   than a certain date, or that match a given tag. This is disabled by default
   and it's under the `delete.garbage_collector` option.

You can read more about this
[here](http://port.us.org/docs/Configuring-Portus.html#delete-support).

LDAP has also seen some updates. First of all, this release includes the changes
described in the `2.3.3` release when it comes to encryption, but it also adds
the new `timeout` option, in which you can tune the timeout in seconds for LDAP
lookups. You can read more about this
[here](http://port.us.org/docs/Configuring-Portus.html#ldap-support).

We have also expanded the `user_permission` section, so administrators can
further tune what regular users can do. In more details:

- We have added the `create_webhook` and the `manage_webhook` options, in order
  to restrict webhook management (it is not restricted by default).
- We have added the `push_images` option, which accepts three possible values
  under its `policy` key:
  - `allow-teams`: the default policy, which works as how Portus used to work up
    until now: owners and contributors of teams can push.
  - `allow-personal`: team policy is removed, non-admin users will only be able
    to push into their personal namespaces.
  - `admin-only`: only administrators are allowed to push images.

You can read a summary of the `user_permission.push_images` option
[here](http://port.us.org/features/3_teams_namespaces_and_users.html#summary-with-all-the-options).

Furthermore, you can now also tune the `pagination` rule being applied to all UI
elements which contain a list (e.g. the list of repositories).

Last but not least, we have increased the default value for the JWT token
expiration time, since it has been reported that the default value was just too
small.

#### Moved portusctl into another project

The `portusctl` tool has been rewritten and moved into its own
[project](https://github.com/openSUSE/portusctl). This has allowed us to expand
its possibilities, since now it will mainly interact with your Portus instance
through the API. The interface of this tool has changed quite a lot, but we
kept the ability to execute commands inside of your Portus instance (i.e. the
existing `exec` command). This new tool is already included in Docker images
based on this `2.4` version of Portus.

#### Changes on the API

We have added new endpoints, as you will see on the list below. We would like to
highlight the `bootstrap` endpoint. This endpoint allows an administrator of a
Portus instance to create the first admin user of Portus and to fetch an
application token that has been created for this same user. This way, you no
longer need the UI in order to perform the first steps of your instance.

Besides this, the Portus UI itself is using more and more this API, instead of
using a more traditional approach. Last but not least, we have changed existing
endpoints with more refined status codes, better response objects, etc. Make
sure to visit the [API documentation](http://port.us.org/docs/API.html).

#### Added bots

We have introduced a new concept: bots. Bots are regular users that are created
by administrators, but with some subtleties:

- A bot doesn't own a personal namespace.
- A bot cannot login via web.
- A bot can only log in with application tokens (a token is generated
  automatically when creating a bot).

#### Namespace deletion

After much delay, we have implemented namespace deletion. You don't have to
change anything from your configuration in order to have this enabled (it
depends on the same `delete.enabled` configuration).

### Features

- 20c7e04acdfb Local login form can be disabled (#1603)
- 23e455516156 webhooks: added the name column (#1581)
- 53ce8967395b Allow contributors to delete repositories/tags (#1696)
- 0c0d46b67f6c config: added option to generally disable push access to non-admin users (#1705)
- f0432d102b49 api: added bootstrap endpoint (#1681)
- 2fcbeda7edbb api: added update methods for namespaces and teams (#1794)
- c9e32324a848 Added permissions on webhooks (#1806)
- bf4e913552b8 config: removed the deprecated `ldap.method` key (#1821)
- 7f82a44078c3 Added the possibility to create bots (#1856)
- 5ee93c4b1b96 background: implemented garbage collector (#1864)
- 94f78377e699 Implemented namespace deletion (#1938)

### Fixes

- 0b7a651e8114 api: take the relative url root into account (#1610)

packaging/suse/portusctl/lib/cli.rb
# frozen_string_literal: true

# :nocov:

# Class implementing the cli interface of portusctl
class Cli < Thor
  check_unknown_options!

  desc "setup", "Configure Portus"
  option "secure", desc: "Toggle SSL usage for Portus", type: :boolean, default: true

  # SSL certificate options
  option "ssl-gen-self-signed-certs",
         desc:    "Generate self-signed certificates",
         type:    :boolean,
         default: false
  option "ssl-certs-dir",
         desc:      "Location of own certificates",
         default:   "",
         long_desc: <<~LONGDESC
           Looks for the following required certificate files in the specified folder:
              * `<custom dir>/<hostname>-ca.key`: the certificate key
              * `<custom dir>/<hostname>-ca.crt`: the certificate file
  LONGDESC
  option "ssl-organization",
         desc:    "SSL certificate: organization",
         default: "SUSE Linux GmbH" # gensslcert -o
  option "ssl-organization-unit",
         desc:    "SSL certificate: organizational unit",
         default: "SUSE Portus example" # gensslcert -u
  option "ssl-email",
         desc:    "SSL certificate: email address of webmaster",
         default: "kontact-de@novell.com" # gensslcert -e
  option "ssl-country",
         desc:    "SSL certificate: country (two letters)",
         default: "DE" # gensslcert -c
  option "ssl-city",
         desc:    "SSL certificate: city",
         default: "Nuernberg" # gensslcert -l
  option "ssl-state",
         desc:    "SSL certificate: state",
         default: "Bayern" # gensslcert -s

  # DB options
  option "db-host", desc: "Database: host", default: "localhost"
  option "db-username", desc: "Database: username", default: "portus"
  option "db-password", desc: "Database: password", default: "portus"
  option "db-name", desc: "Database: name", default: "portus_production"

  # Registry
  option "local-registry", desc: "Configure Docker registry running locally",
    type: :boolean, default: false

  # LDAP
  option "ldap-enable", desc: "LDAP: enable", type: :boolean, default: false
  option "ldap-hostname", desc: "LDAP: server hostname"
  option "ldap-port", desc: "LDAP: server port", default: "389"
  option "ldap-method",
         desc:    "LDAP: encryption method",
         default: "plain"
  option "ldap-encryption-method",
         desc:    "LDAP: encryption method",
         default: "plain"
  option "ldap-encryption-options-ca-file",
         desc:    "LDAP: path to the CA file to be passed",
         default: ""
  option "ldap-encryption-options-ssl-version",
         desc:    "LDAP: SSL version to be used for communications",
         default: "TLSv1_2"
  option "ldap-base", desc: "LDAP: base", default: "ou=users, dc=example, dc=com"
  option "ldap-filter", desc: "LDAP: filter users"
  option "ldap-uid", desc: "LDAP: uid", default: "uid"
  option "ldap-authentication-enable",
         desc:    "LDAP: enable LDAP credentials for user lookup",
         type:    :boolean,
         default: false
  option "ldap-authentication-bind-dn", desc: "LDAP: bind DN for authentication"
  option "ldap-authentication-password", desc: "LDAP: password for authentication"
  option "ldap-guess-email-enable",
         desc:    "LDAP: guess email address",
         type:    :boolean,
         default: false
  option "ldap-guess-email-attr",
         desc: "LDAP: attribute to use when guessing email address"

  # OAuth
  require_relative "cli_oauth"
  include ::CliOAuth

  # MAILER
  option "email-from",
         desc:    "MAIL: sender address",
         default: "portus@#{HOSTNAME}"
  option "email-name", desc: "MAIL: sender name", default: "Portus"
  option "email-reply-to",
         desc:    "MAIL: reply to address",
         default: "no-reply@#{HOSTNAME}"
  option "email-smtp-enable",
         desc:    "MAIL: use SMTP as the delivery method",
         type:    :boolean,
         default: false
  option "email-smtp-address",
         desc:    "MAIL: the address to the SMTP server",
         default: "smtp.example.com"
  option "email-smtp-port", desc: "MAIL: SMTP server port", default: "587"
  option "email-smtp-username",
         desc:    "MAIL: the user name to be used for logging in the SMTP server",
         default: "username@example.com"
  option "email-smtp-password",
         desc:    "MAIL: the password to be used for logging in the SMTP server",
         default: "password"
  option "email-smtp-domain",
         desc:    "MAIL: the domain of the SMTP server",
         default: "example.com"

  # SIGNUP
  option "signup-enable",
         desc:    "Enable user signup",
         type:    :boolean,
         default: true

  # GRAVATAR
  option "gravatar-enable",
         desc:    "Enable Gravatar usage",
         type:    :boolean,
         default: true

  # JWT EXPIRATION TIME
  option "jwt-expiration-time",
         desc:    "Expiration time for the JWT token used by Portus",
         type:    :numeric,
         default: 5

  # Catalog pagination
  option "catalog-page",
         desc:    "Pagination value for API calls to the registry",
         type:    :numeric,
         default: 100

  option "registry-timeout-value",
         desc:    "Timeout value for API calls to the registry",
         type:    :numeric,
         default: 10

  option "registry-read-timeout-value",
         desc:    "Read timeout value for API calls to the registry",
         type:    :numeric,
         default: 10

  # FIRST USER
  option "first-user-admin-enable",
         desc:    "Make the first registered user an admin",
         type:    :boolean,
         default: true

  # Display name
  option "display-name-enable",
         desc:    "Enable users to set a display name",
         type:    :boolean,
         default: false

  option "delete-enable",
         desc:    "Enable delete support. Only do this if your registry is 2.4 or higher",
         type:    :boolean,
         default: false

  option "change-visibility-enable",
         desc:    "Allow users to change the visibility of their namespaces",
         type:    :boolean,
         default: true

  option "manage-namespace-enable",
         desc:    "Allow users to modify their namespaces",
         type:    :boolean,
         default: true

  option "create-namespace-enable",
         desc:    "Allow users to modify new namespaces",
         type:    :boolean,
         default: true

  option "manage-team-enable",
         desc:    "Allow users to modify their teams",
         type:    :boolean,
         default: true

  option "create-team-enable",
         desc:    "Allow users to create new teams",
         type:    :boolean,
         default: true

  # Security scanning
  option "security-clair-server",
         desc:    "The URL allowing Portus to access your CoreOS Clair server",
         default: ""

  option "security-clair-health-port",
         desc:    "The port in which Clair exposes the /health endpoint",
         type:    :numeric,
         default: 6061

  option "security-clair-timeout",
         desc:    "The timeout for HTTP requests targeting Clair",
         type:    :numeric,
         default: 900

  option "security-zypper-server",
         desc:    "The URL allowing Portus to access your zypper-docker server",
         default: ""

  option "security-dummy-server",
         desc:    "If non-empty, then Portus will fake a 'dummy' server (only for development)",
         default: ""

  # Anonymous browsing
  option "anonymous-browsing-enable",
         desc:    "Allow anonymous users to explore public repositories",
         type:    :boolean,
         default: true

  option "background-registry-enable",
         desc:    "Toggle registry support",
         type:    :boolean,
         default: true

  option "background-sync-enable",
         desc:    "Toggle synchronization support",
         type:    :boolean,
         default: true

  option "background-sync-strategy",
         desc:    "Set the synchronization strategy for the sync task",
         default: "initial"

  def setup
    ensure_root
    check_setup_flags options

    configure = Configurator.new(options)
    configure.apache
    configure.ssl
    configure.database_config
    registry_config = configure.registry
    configure.secrets
    configure.config_local
    configure.create_database
    configure.services

    return if options["local-registry"]

    puts "Ensure the registry running on another host is configured properly"
    puts "This is a working configuration file you might want to use:"
    puts registry_config
  end

  desc "make_admin USERNAME", "Give 'admin' role to a user"
  def make_admin(username)
    if username.blank?
      # This will print the list of usernames
      Runner.bundler_exec("rake", "portus:make_admin", {})
    else
      # Rake tasks look weird when they accept parameters
      Runner.bundler_exec("rake", "portus:make_admin[#{username}]", {})
    end
  end

  desc "rake ARGS...", "Run a rake task against Portus"
  def rake(*args)
    if args.empty?
      warn "You mush provide at least an argument"
      exit 1
    end

    Runner.bundler_exec("rake", args, {})
  end

  desc "exec ARGS...", "Run a arbitrary command via bundler exec"
  def exec(*args)
    if args.empty?
      warn "You mush provide at least an argument"
      exit 1
    end
    exec_args = []
    exec_args = args[1, args.size] if args.size > 1

    Runner.bundler_exec(args[0], exec_args, {})
  end

  desc "logs", "Collect all the logs used for debugging purposes"
  def logs(*args)
    warn "Extra arguments ignored..." unless args.empty?
    ensure_root

    Runner.produce_versions_file!
    Runner.exec("cp", ["/var/log/apache2/error_log", File.join(PORTUS_ROOT, "log/production.log")])
    Runner.tar_files("log/production.log", "log/versions.log")
  end
end

# :nocov:
